dkg/pedersen/mod.go

```diff
@@ -65,41 +68,50 @@ type Actor struct {
 }
 
 // Setup implement dkg.Actor. It initializes the DKG.
-func (a *Actor) Setup(players mino.Players, pubKeys []kyber.Point, threshold int) error {
-
-	if players.Len() != len(pubKeys) {
-		return xerrors.Errorf("there should be as many players as "+
-			"pubKey: %d := %d", players.Len(), len(pubKeys))
-	}
+func (a *Actor) Setup(co crypto.CollectiveAuthority, threshold int) (kyber.Point, error) {
 
 	if a.startRes.Done() {
-		return xerrors.Errorf("startRes is already done, only one setup call is allowed")
+		return nil, xerrors.Errorf("startRes is already done, only one setup call is allowed")
 	}
 
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
 
-	sender, receiver, err := a.rpc.Stream(ctx, players)
+	sender, receiver, err := a.rpc.Stream(ctx, co)
 	if err != nil {
-		return xerrors.Errorf("failed to stream: %v", err)
+		return nil, xerrors.Errorf("failed to stream: %v", err)
 	}
 
-	players.AddressIterator().Seek(0)
-	addrs := make([]mino.Address, 0, players.Len())
-	for players.AddressIterator().HasNext() {
-		addrs = append(addrs, players.AddressIterator().GetNext())
+	co.AddressIterator().Seek(0)
+	addrs := make([]mino.Address, 0, co.Len())
+	pubkeys := make([]kyber.Point, 0, co.Len())
+	iter := co.AddressIterator()
+	for iter.HasNext() {
+		addr := iter.GetNext()
+		addrs = append(addrs, addr)
+		pubkey, index := co.GetPublicKey(addr)
```

Comment:
Use two iterators: `co.AddressIterator()` and `co.PublicKeyIterator()` so you can avoid the search. Then you can remove the `PubKeyNotFound` thing. The implementation of the collective authority is responsible for the iterators to be consistent.

Reply:
Actually even better: why don't you have the collective authority as an attribute ? So you don't need to transform stuff into arrays (I mean eventually yes, but it's implemented in the serialization logic, not here).

Reply:
I need the list of addresses anyway to call sender.send(). Since I can't fully skip transforming stuff into arrays I don't think its worth changing the attribute (plus it would just move the problem since at the end I still need those slices).

Reply:
Ah yes, that's something to think about for the mino interface.